### ISSUE TITLE
Informative exceptions

### ISSUE THREAD

**Issue by treeowl** [conal/unamb#5]
When `x` and `y` both throw exceptions, `unamb x y` throws a totally uninformative exception. It would be very nice to have a version that throws both exceptions. That is, that throws something like

```haskell
data InformativeBottoms = forall e1 e2. (Exception e1, Exception e2) => InformativeBottoms e1 e2
  deriving Show
instance Exception InformativeBottoms
```

I imagine this is sometimes expensive, so perhaps there should be a separate version of `unamb` that preserves exceptions like this.

What led me here:

I was playing around with `lub`, and realized that mismatched constructors don't give informative errors. For example, if I write

```haskell
lub (expensiveThing `seq` Left 3) (Right undefined)
```

then I don't get a useful error message. The system *could* provide one, in principle. We commit to `Right` (because it takes a long time to get to the `Left` constructor), but we have to wait on the `Left` computation to (try) to get the payload. So when we match on the `Left` constructor, we already know it's wrong. But we match on it under the `unamb` to calculate the `Integer`, so if we throw out a mismatch error it gets swallowed by `BothBottom`. *sigh*.
